File: 99.other/nifdc-crawler/js/nifdc.js
const _ = require('lodash');
const superagent = require('superagent');
require('superagent-charset')(superagent);
const cheerio = require('cheerio');

const SQ_COL_NAME = '序号';
const AMOUNT_OLD_COL_NAME = '批量/进口量';
const AMOUNT_VALUE_COL_NAME = '数量';
const AMOUNT_UNIT_COL_NAME = '单位';
const SHEET_NAME = 'NAME';

//excel列字母编号转为数字（以1开始）
function colString2int(str) {
    let col = 0;
    for (let i = str.length - 1; i >= 0; i--) {
        col += (str.charCodeAt(str.length - i - 1) - 64) * Math.pow(26, i);
    }
    return col;
}

//excel列数字序号转为字母编号（以1开始）
function int2colString(num) {
    let charArray = [];
    let n = num;
    while (n >= 1) {
        let a = n % 26;
        charArray.unshift(String.fromCharCode((a == 0 ? 26 : a) + 64));
        n = Math.floor(n / 26);
        if (a == 0) {
            n--;
        }
    }
    return charArray.join('');
}

function getSheetName(title) {
    let lIdx = title.lastIndexOf("（");
    let rIdx = title.lastIndexOf("）");
    return _.trim(title.substr(lIdx + 1, rIdx - lIdx - 1));
}

function isSpecialCharacter(code) {
    if (code <= 0x206F && code >= 0x2000) {
        return true;
    }
    return false;
}

//判断行是否存在数据
function hasData(r) {
    for (var i = 0; i < r.length; i++) {
        let v = r[i];
        if (v && !isSpecialCharacter(v.codePointAt(0))) {
            return true;
        }
    }
    return false;
}

async function crawlItem(itemUrl) {

    let itemTable = {};
    await superagent.get(itemUrl)
        .charset('gbk')
        .then(function (sres) {
            console.log(`end get ${itemUrl}`);
            let $ = cheerio.load(sres.text);
            let rowspans = [];
            let title = $('.articletitle1').text();
            itemTable['title'] = getSheetName(title);
            let data = [];
            let tbs = $('.articlecontent1 table');
            if (tbs.length > 0) {
                let tb = $(tbs[tbs.length - 1]);
                $(tb).find('tr').each(function (idxRow, element) {
                    let item = {};
                    let colOffset = 0;
                    let row = [];
                    $(element).find('td').each(function (idxCol, el) {
                        let idx = idxCol + colOffset;
                        if (rowspans[idx] > 1) {
                            row.push(data[idxRow - 1][idx]);
                            colOffset++;
                            rowspans[idx] = rowspans[idx] - 1;
                            idx++;
                        }
                        let $el = $(el);
                        rowspan = parseInt($el.attr('rowspan'));
                        if (isNaN(rowspan)) {
                            rowspans[idx] = 1;
                        } else {
                            rowspans[idx] = parseInt(rowspan);
                        }
                        let txt = $el.text();
                        txt = _.trim(txt);
                        row.push(txt);
                    });
                    //跳过内容为空的行
                    if (hasData(row)) {
                        data.push(row);
                    }
                });
            }

            itemTable.data = data;
        });

    return itemTable;
}

async function crawlItems(items) {
    let oItems = _.sortBy(items, ['title']);
    let allItemData = [];

    for (let i = 0; i < oItems.length; i++) {
        let item = oItems[i];
        let itemTable = await crawlItem(item.itemUrl);
        allItemData = _.concat(allItemData, itemTable.data);
    }
    if (items && items.length > 0 && allItemData.length > 0) {
        //处理标题，去除标题中间空格
        let titleRow = allItemData[0];
        _.forEach(titleRow, function (v, k) {
            titleRow[k] = _.replace(v, ' ', '');
        });
        //增加序号列
        let sq = findColData(titleRow, titleRow, SQ_COL_NAME);
        if (!(sq == SQ_COL_NAME)) { //不存在序号列
            titleRow.push(SQ_COL_NAME);
            for (let idx = 1; idx < allItemData.length; idx++) { //跳过标题行
                let row = allItemData[idx];
                row.push(idx.toString());
            }
        }
        return {
            name: items[0]['name'],
            data: allItemData,
            title: titleRow,
        };
    }
    return null;
}

async function crawlYear(yearUrl, extendItems) {
    let results = [];
    await superagent.get(yearUrl)
        .charset('gbk')
        .then(async function (sres) {
            console.log(`end get ${yearUrl}`);
            let $ = cheerio.load(sres.text);
            let items = [];
            if (extendItems && extendItems.length > 0) {
                _.forEach(extendItems, function (v, k) {
                    items.push({ title: v.title, name: getSheetName(v.title), itemUrl: v.url });
                })
            }
            $('.ListColumnClass1 > a').each(function (idx, el) {
                let $el = $(el);
                let title = $el.text();
                let name = getSheetName(title);
                let itemUrl = yearUrl + $el.attr('href');
                items.push({ title: title, name: name, itemUrl: itemUrl });
            });
            let gItems = _.groupBy(items, 'name');
            let promises = _.map(gItems, function (v) {
                return crawlItems(v);
            });
            results = await Promise.all(promises);
        });
    return _.compact(results);
}

function buildSheet(itemData) {
    let item_sheet = {};
    let allItemData = itemData.data;

    let titleRow = itemData.title;
    item_sheet['name'] = itemData.name;
    item_sheet['!ref'] = 'A1:' + int2colString(titleRow.length) + (allItemData.length);
    item_sheet['!cols'] = [{ wch: 6 }];
    for (let i = 0; i < allItemData.length; i++) {
        let row = allItemData[i];
        for (let j = 0; j < row.length; j++) {
            let col = row[j];
            const c = int2colString(j + 1);
            item_sheet[c + (i + 1)] = { v: col };
        }
    }

    return item_sheet;
}

//查找对应列数据
function findColData(titleRow, r, title) {
    let idx = _.findIndex(titleRow, function (o) { return o == title; });
    if (idx >= 0) {
        return r[idx];
    }
    return null;
}

//判断是否为数据行,通过序号是否为数字进行判断
function isDataRow(titleRow, r) {
    let sq = findColData(titleRow, r, SQ_COL_NAME);
    if (isNaN(parseInt(sq))) {
        return false;
    }
    return true;
}

function buildAllSheet(itemDatas) {
    let all_sheet = { name: 'ALL' };
    let all_title = _.concat(SHEET_NAME, _.union.apply(this, _.map(itemDatas, 'title')));
    let idxAllRow = 1;
    //合并后的标题行
    all_title.push(AMOUNT_VALUE_COL_NAME);
    all_title.push(AMOUNT_UNIT_COL_NAME);
    _.forEach(all_title, function (title, idxCol) {
        const c = int2colString(idxCol + 1);
        all_sheet[c + idxAllRow] = { v: title };
    });
    idxAllRow++;
    //循环itemDatas，合并数据
    _.forEach(itemDatas, function (item) {
        //循环行
        _.forEach(item.data, function (r, idxRow) {
            if (isDataRow(item.title, r)) { //仅合并数据行
                let amount = findColData(item.title, r, AMOUNT_OLD_COL_NAME);
                //循环列
                _.forEach(all_title, function (title, idxCol) {
                    const c = int2colString(idxCol + 1);
                    if (_.isEqual(title, SHEET_NAME)) { //表格名
                        all_sheet[c + idxAllRow] = { v: item.name };
                    } else if (_.isEqual(title, AMOUNT_OLD_COL_NAME)) {
                        all_sheet[c + idxAllRow] = { v: amount ? amount : '' };
                    } else if (_.isEqual(title, AMOUNT_VALUE_COL_NAME)) {
                        let aValue = amount.match(/\d+/);
                        let av = aValue ? aValue[0] : 0;
                        all_sheet[c + idxAllRow] = { t: 'n', v: av };
                    } else if (_.isEqual(title, AMOUNT_UNIT_COL_NAME)) {
                        let aUnit = amount.match(/\D+/);
                        let au = aUnit ? aUnit[0] : '';
                        all_sheet[c + idxAllRow] = { v: au };
                    } else {
                        let d = findColData(item.title, r, title);
                        all_sheet[c + idxAllRow] = { v: d ? d : '' };
                    }
                });
                idxAllRow++;
            }
        });
    })

    all_sheet['!ref'] = 'A1:' + int2colString(all_title.length) + idxAllRow;
    all_sheet['!cols'] = [{ wch: 24 }, { wch: 4 }];

    return all_sheet;
}

function buildWorkbook(itemDatas) {
    let out_workbook = {};

    let sheets = {};
    _.forEach(itemDatas, function (itemData) {
        let sheet = buildSheet(itemData);
        sheets[sheet['name']] = sheet;
    });
    let all_sheet = buildAllSheet(itemDatas);
    out_workbook = {
        SheetNames: _.concat(all_sheet.name, _.keys(sheets)),
        Sheets: sheets
    };
    sheets[all_sheet['name']] = all_sheet;

    return out_workbook;
}

async function createYearWorkbook(yearUrl, extendItems) {
    let itemDatas = await crawlYear(yearUrl, extendItems);
    return buildWorkbook(itemDatas);
}

exports.createYearWorkbook = createYearWorkbook;

File: 12.go/src/nifdc-crawler/nifdc.go
package main

import (
	"sort"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"regexp"

	"github.com/PuerkitoBio/goquery"
	"golang.org/x/net/html/charset"
	"github.com/tealeg/xlsx"
)

const SQ_COL_NAME = "序号"
const AMOUNT_OLD_COL_NAME = "批量/进口量"
const AMOUNT_VALUE_COL_NAME = "数量"
const AMOUNT_UNIT_COL_NAME = "单位"
const SHEET_NAME = "NAME"

type NifdcConfigExtend struct {
	Name string `json:"name,omitempty"`
	Title string `json:"title,omitempty"`
	URL   string `json:"url,omitempty"`
}

type ExtendItemSlice []NifdcConfigExtend
func (s ExtendItemSlice) Len() int { return len(s) }
func (s ExtendItemSlice) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
func (s ExtendItemSlice) Less(i, j int) bool { return s[i].Title < s[j].Title }

type NifdcConfigInfo struct {
	Enable      bool                `json:"enable,omitempty"`
	Year        string              `json:"year,omitempty"`
	URL         string              `json:"url,omitempty"`
	ExtendItems ExtendItemSlice `json:"extendItems,omitempty"`
}

type NifdcConfig struct {
	DataInfo []NifdcConfigInfo `json:"dataInfo,omitempty"`
}

type DataRow = []string

type DataTable struct {
	Name string
	Title string
	TitelRow DataRow
	Data  []DataRow
}

type Nifdc struct {
	config *NifdcConfig
	logFunc func(msg string)
}

func (n *Nifdc) init(config *NifdcConfig) {
	n.config = config
}

func (n *Nifdc) crawlItem(itemUrl string) *DataTable {
	res, err := http.Get(itemUrl)
	if err != nil {
		fmt.Println(err.Error())
	} else {
		defer res.Body.Close()
		utfBody, err := charset.NewReader(res.Body, "gb18030")
		if err != nil {
			fmt.Println(err.Error())
		} else {
			doc, _ := goquery.NewDocumentFromReader(utfBody)
			dt := new(DataTable)
			dt.Title = doc.Find(".articletitle1").Text()
			dt.Name = n.getSheetName(dt.Title)
			var rowspans []int
			tb := doc.Find(".articlecontent1 table").Last()
			tb.Find("tr").Each(func(idxRow int, r *goquery.Selection) {

				colOffset := 0
				var row DataRow
				r.Find("td").Each(func(idxCol int, c *goquery.Selection) {
					idx := idxCol + colOffset
					if len(rowspans) <= idxCol {
						rowspans = append(rowspans, 1)
					}
					if rowspans[idx] > 1 {
						row = append(row, dt.Data[idxRow-1][idx])
						colOffset++
						rowspans[idx] = rowspans[idx] - 1
						idx++
					}
					sRowspan, _ := c.Attr("rowspan")
					rowspan, err := strconv.Atoi(sRowspan)
					if err != nil {
						rowspans[idx] = 1
					} else {
						rowspans[idx] = rowspan
					}
					row = append(row, strings.TrimSpace(c.Text()))

				})
				//跳过内容为空的行
				if (n.hasData(&row)) {
					dt.Data = append(dt.Data, row)
				}
			})

			return dt
		}
	}
	return nil
}

func (n *Nifdc) crawlItems(items ExtendItemSlice) *DataTable{
	sort.Stable(items)
    oItems := items
    var allItemData []DataRow

    for i := range oItems {
        item := oItems[i];
		n.log(fmt.Sprintf("fetch %s %s",item.Title,item.URL))
        itemTable := n.crawlItem(item.URL);
        allItemData = append(allItemData, itemTable.Data...);
    }
    if items != nil && len(items) > 0 && len(allItemData) > 0 {
        //处理标题，去除标题中间空格
		titleRow := allItemData[0];
		for k,v := range titleRow {
			titleRow[k] = strings.Replace(v," ","",-1)
		}
        //增加序号列
        sq := n.findColData(&titleRow, &titleRow, SQ_COL_NAME);
        if !(sq == SQ_COL_NAME) { //不存在序号列
            titleRow = append(titleRow,SQ_COL_NAME);
            for  idx := 1; idx < len(allItemData); idx++ { //跳过标题行
                row := allItemData[idx];
                row = append(row,strconv.Itoa(idx));
            }
		}
		dt := new(DataTable)
		dt.Data = allItemData
		dt.TitelRow = titleRow
		dt.Name = items[0].Name
		return dt
    }
    return nil;
}

func (n *Nifdc) crawlYear(info NifdcConfigInfo) []DataTable{
	var results []DataTable
	
	res, err := http.Get(info.URL)
	if err != nil {
		fmt.Println(err.Error())
	} else {
		defer res.Body.Close()
		utfBody, err := charset.NewReader(res.Body, "gb18030")
		if err != nil {
			fmt.Println(err.Error())
		} else {
			doc, _ := goquery.NewDocumentFromReader(utfBody)
			var items ExtendItemSlice
            if (info.ExtendItems != nil && len(info.ExtendItems) > 0) {
				for i := 0; i < len(info.ExtendItems); i++ {
					item := info.ExtendItems[i]
					item.Name = n.getSheetName(item.Title)
				}
				items = append(items,info.ExtendItems...)
			}
			doc.Find(".ListColumnClass1 > a").Each(func(idx int, el *goquery.Selection) {
				var item NifdcConfigExtend
				href,_ := el.Attr("href")
				item.Title = el.Text()
				item.URL = info.URL + href
				item.Name = n.getSheetName(item.Title)
				items = append(items,item)
			})
			allTitle := make([]string, 0, len(items))
			gItems := make(map[string]ExtendItemSlice)
			for _, v := range items {
				if val, ok := gItems[v.Name]; ok {
					gItems[v.Name] = append(val, v)
				} else{
					gItems[v.Name] = ExtendItemSlice{v}
					allTitle = append(allTitle, v.Name)
				}
			}
			
			for _,v := range allTitle{
				dt := n.crawlItems(gItems[v])
				results = append(results,*dt)
			}
		}
	}
	return results
}

func (n *Nifdc) getSheetName(title string) string {

	lIdx := strings.LastIndex(title, "（")
	rIdx := strings.LastIndex(title, "）")
	return strings.TrimSpace(title[lIdx + len("（") : rIdx]) //（占3个字节
}

func (n *Nifdc) isSpecialCharacter(code rune) bool {
	if code <= 0x206F && code >= 0x2000 {
		return true
	}
	return false
}

//判断行是否存在数据
func (n *Nifdc) hasData(r *DataRow) bool {
	for i := range *r {
		v := (*r)[i]
		if len(v) > 0 && !n.isSpecialCharacter(rune(v[0])) {
			return true
		}
	}
	return false
}

//查找对应列数据
func (n *Nifdc) findColData(titleRow *DataRow, r *DataRow, title string) string {
	idx:= -1
	for i := range *titleRow {
		if (*titleRow)[i] == title{
			idx = i;
			break
		}
	}
    if (idx >= 0) {
        return (*r)[idx];
    }
    return "";
}

//判断是否为数据行,通过序号是否为数字进行判断
func (n *Nifdc) isDataRow(titleRow *DataRow, r *DataRow) bool{
	sq := n.findColData(titleRow, r, SQ_COL_NAME);
	_, err := strconv.Atoi(sq)
    if (err != nil) {
        return false;
    }
    return true;
}

func (n *Nifdc) buildSheet(itemData *DataTable) *xlsx.Sheet{
	sheet := new(xlsx.Sheet)
	sheet.Name = itemData.Name
	for _,r := range itemData.Data{
		row := sheet.AddRow()
		for _,c := range r{
			cell:= row.AddCell()
			cell.Value = c
		}
	}
    return sheet;
}

func (n *Nifdc) buildAllSheet(itemDatas []DataTable)  *xlsx.Sheet{
	allSheet := new(xlsx.Sheet)
	allSheet.Name = "ALL"

	allTitle := make([]string, 0, len(itemDatas[0].TitelRow))
	allTitle = append(allTitle, SHEET_NAME)
	allTitleMap := make(map[string]bool)
	for _,table := range itemDatas {
		for _,t := range table.TitelRow {
			if _, ok := allTitleMap[t]; !ok {
				allTitleMap[t] = true
				allTitle = append(allTitle, t)
			}
		}
	}

	//合并后的标题行
	allTitle = append(allTitle,AMOUNT_VALUE_COL_NAME)
	allTitle = append(allTitle,AMOUNT_UNIT_COL_NAME)
	titleRow := allSheet.AddRow()
	for _,t := range allTitle {
		c := titleRow.AddCell()
		c.Value = t
	}

	//循环itemDatas，合并数据
	for _,item := range itemDatas {
        //循环行
		for _,r := range item.Data {
			if n.isDataRow(&item.TitelRow, &r) { //仅合并数据行
				row := allSheet.AddRow()
				amount := n.findColData(&item.TitelRow, &r, AMOUNT_OLD_COL_NAME);
				//循环列
				for _,title := range allTitle {
					cell := row.AddCell()
					cell.Value = ""
					if title == SHEET_NAME { //表格名
						cell.Value = item.Name
                    } else if title == AMOUNT_OLD_COL_NAME {
							cell.Value = amount
                    } else if title == AMOUNT_VALUE_COL_NAME {
						aValue := regexp.MustCompile(`\d+`).FindAllString(amount,-1)
						if aValue != nil && len(aValue) > 0 {
							cell.Value = aValue[0] //数字
						}
                    } else if title == AMOUNT_UNIT_COL_NAME {
						aUnit := regexp.MustCompile(`\D+`).FindAllString(amount,-1)
						if aUnit != nil && len(aUnit) > 0 {
							cell.Value = aUnit[0]
						}
                    } else {
                        d := n.findColData(&item.TitelRow, &r, title);
							cell.Value = d
                    }
                }
            }
		}
	}

    return allSheet;
}

func (n *Nifdc) buildWorkbook(itemDatas []DataTable) *xlsx.File{
    workbook := xlsx.NewFile()

	allSheet := n.buildAllSheet(itemDatas);
	workbook.AppendSheet(*allSheet,allSheet.Name)
	for _,v := range itemDatas{
		sheet := n.buildSheet(&v)
		workbook.AppendSheet(*sheet,sheet.Name)
	}

    return workbook;
}

func (n *Nifdc) createYearWorkbook(info NifdcConfigInfo)  *xlsx.File{
    itemDatas := n.crawlYear(info);
    return n.buildWorkbook(itemDatas);
}

func (n *Nifdc) log(msg string){
	if n.logFunc != nil{
		n.logFunc(msg)
	}
}